File: app/src/main/java/com/frank/ffmpeg/activity/MediaPlayerActivity.java
package com.frank.ffmpeg.activity;

import android.os.Bundle;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;

import com.frank.ffmpeg.MediaPlayer;
import com.frank.ffmpeg.R;
import com.frank.ffmpeg.util.FileUtil;

/**
 * mediaPlayer, which decode by software
 * Created by frank on 2018/2/12.
 */

public class MediaPlayerActivity extends BaseActivity implements SurfaceHolder.Callback {

    private static final String TAG = MediaPlayerActivity.class.getSimpleName();

    private SurfaceHolder surfaceHolder;

    private MediaPlayer mediaPlayer;

    private boolean surfaceCreated;

    private Button btnSelectFile;

    @Override
    int getLayoutId() {
        return R.layout.activity_media_player;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        hideActionBar();
        initView();
        initPlayer();
    }

    private void initView(){
        SurfaceView surfaceView = getView(R.id.surface_media);
        surfaceHolder = surfaceView.getHolder();
        surfaceHolder.addCallback(this);
        btnSelectFile = getView(R.id.btn_select_file);
        initViewsWithClick(R.id.btn_select_file);
    }

    private void initPlayer(){
        mediaPlayer = new MediaPlayer();
    }

    @Override
    public void surfaceCreated(SurfaceHolder holder) {
        surfaceCreated = true;
    }

    @Override
    public void surfaceChanged(SurfaceHolder holder, int format, int width, int height) {
        Log.i(TAG, "surfaceChanged");
    }

    @Override
    public void surfaceDestroyed(SurfaceHolder holder) {
        Log.i(TAG, "surfaceDestroyed");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if(mediaPlayer != null){
            mediaPlayer.release();
            mediaPlayer = null;
        }
    }

    private void startPlay(final String filePath) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                int result = mediaPlayer.setup(filePath, surfaceHolder.getSurface());
                if(result < 0){
                    Log.e(TAG, "mediaPlayer setup error!");
                    return;
                }
                mediaPlayer.play();
            }
        }).start();
    }

    @Override
    void onSelectedFile(String filePath) {
        if (!FileUtil.checkFileExist(filePath)){
            return;
        }
        if (surfaceCreated) {
            btnSelectFile.setVisibility(View.GONE);
            startPlay(filePath);
        }
    }

    @Override
    void onViewClick(View view) {
        if (view.getId() == R.id.btn_select_file) {
            selectFile();
        }
    }
}

File: app/src/main/java/com/frank/ffmpeg/activity/MediaPlayerActivity.kt
package com.frank.ffmpeg.activity

import android.os.Bundle
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.View
import android.widget.Button

import com.frank.ffmpeg.MediaPlayer
import com.frank.ffmpeg.R
import com.frank.ffmpeg.util.FileUtil

/**
 * mediaPlayer, which decode by software
 * Created by frank on 2018/2/12.
 */

class MediaPlayerActivity : BaseActivity(), SurfaceHolder.Callback {

    private var surfaceHolder: SurfaceHolder? = null

    private var mediaPlayer: MediaPlayer? = null

    private var surfaceCreated: Boolean = false

    private var btnSelectFile: Button? = null

    override val layoutId: Int
        get() = R.layout.activity_media_player

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        hideActionBar()
        initView()
        initPlayer()
    }

    private fun initView() {
        val surfaceView = getView<SurfaceView>(R.id.surface_media)
        surfaceHolder = surfaceView.holder
        surfaceHolder!!.addCallback(this)
        btnSelectFile = getView(R.id.btn_select_file)
        initViewsWithClick(R.id.btn_select_file)
    }

    private fun initPlayer() {
        mediaPlayer = MediaPlayer()
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        surfaceCreated = true
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        Log.i(TAG, "surfaceChanged")
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        Log.i(TAG, "surfaceDestroyed")
    }

    override fun onDestroy() {
        super.onDestroy()
        if (mediaPlayer != null) {
            mediaPlayer!!.release()
            mediaPlayer = null
        }
    }

    private fun startPlay(filePath: String) {
        Thread(Runnable {
            val result = mediaPlayer!!.setup(filePath, surfaceHolder!!.surface)
            if (result < 0) {
                Log.e(TAG, "mediaPlayer setup error!")
                return@Runnable
            }
            mediaPlayer!!.play()
        }).start()
    }

    override fun onSelectedFile(filePath: String) {
        if (!FileUtil.checkFileExist(filePath)) {
            return
        }
        if (surfaceCreated) {
            btnSelectFile!!.visibility = View.GONE
            startPlay(filePath)
        }
    }

    override fun onViewClick(view: View) {
        if (view.id == R.id.btn_select_file) {
            selectFile()
        }
    }

    companion object {

        private val TAG = MediaPlayerActivity::class.java.simpleName
    }
}
